### ISSUE TITLE
double free in tcache2 or segfault when using Vec in async

### ISSUE THREAD

**Issue by SiebenCorgie** [rust-lang/rust#75117]
Hi,

I recently started to try my luck with async code via [smol](https://crates.io/crates/smol). However, I a strange bug when using vectors in smols async context, where my program crashes with 

`free(): double free detected in tcache 2` or a segfault

This is the code, a description follows below:
```rust
extern crate async_graph;
use async_graph::*;
use std::sync::{Arc, RwLock};


enum SenderEn{
    Something,
    SomeVec(Vec<f32>)
}

struct SenderNode{
    some_inner: i32,
    channel: Vec<Arc<RwLock<dyn Splitter<DataLayout = [SenderEn; 1]> + Send + Sync>>>,
}

impl ProducerAble for SenderNode{
    type OutputLayout = [SenderEn; 1];
    
    fn receiver_channel(&mut self, channel: Arc<RwLock<dyn Splitter<DataLayout = Self::OutputLayout> + Send + Sync>>){
	self.channel.push(channel);
    }

    fn produce(&self){
	for c in &self.channel{
	    async_graph::smol::run(async{
		println!("Works: ");
		c.read().unwrap().split([SenderEn::Something], "Nothing").unwrap();
	    });

// --- Problem below	    
	    async_graph::smol::run(async{
		println!("Works as well: ");
		let mut vec = Vec::with_capacity(10);
		//When removing this print, crash occurs as well but is double free, not segfault
		println!("After alloc");
		for _ in 0..10{
		    vec.push(0.0);
		}
		c.read().unwrap().split([SenderEn::SomeVec(vec)], "Created vec").unwrap();
	    });

	    
	    async_graph::smol::run(async{
		println!("Crashes: "); //crashes with segfault on nightly
		c.read().unwrap().split([SenderEn::SomeVec(vec![0.0; 64])], "Created vec").unwrap();
	    });
	}
    }
}

impl Node for SenderNode{
    type Inputs = ();
    type Outputs = [SenderEn; 1];

    fn process(&self, input: Self::Inputs) -> Self::Outputs{
	[SenderEn::Something]
    }
}


fn main() {    
    //Run producer then wait
    let node = Arc::new(RwLock::new(SenderNode{some_inner: 654, channel: Vec::with_capacity(1)}));
    let prod = Aggregator::from_arc_producer(node.clone());

    let node_clone = node.clone();
    node_clone.read().unwrap().produce();

    std::thread::sleep(std::time::Duration::from_secs(1));
    let oke = node.read().unwrap().some_inner;
}
```

## Description 
I currently try to build some kind of async execution DAG (async_graph). The DAG itself work, but has this bug. A Producer is some `Node` that can send values via `produce` into the graph from some thread (in this case the main thread). However, in this example no nodes are connected, so the output of produce is send nowhere.

The `Aggregator` is a wrapper around some type that implements `Node`. It handles the connection from and to this node as well as splitting input from a Box<Any> into tuples of (X,Y,Z) (based on the associated type of `Node`). Similarly it handles assembling some output of the nodes `process` function into several send calls to other nodes.

The [async_graph](https://gitlab.com/tendsinmende/asyncgraph) is open source and documented if more information regarding it is needed.

So in short, something goes wrong when working with vecs in my special, async case. I do not use any unsafe code, so I think this shouldn't happen.

### Meta
I tested it on nightly.

`rustc --version --verbose`:
```
rustc 1.47.0-nightly (6c8927b0c 2020-07-26)
binary: rustc
commit-hash: 6c8927b0cf80ceee19386026cf9d7fd4fd9d486f
commit-date: 2020-07-26
host: x86_64-unknown-linux-gnu
release: 1.47.0-nightly
LLVM version: 10.0
```
Hardware / Software
OS: Updated arch linux
CPU: AMD Ryzen 1700x

## backtrace
Sadly there is no backtrace.

## Closing
I uploaded the project which can be executed and should produce the same problem: https://gitlab.com/tendsinmende/async_double_free

If there is anything I can debug/do to help, let me know!

Cheers
Tendsin Mende

**Comment by jonas-schievink** [comment]
Please make sure this is a bug in Rust and not in the unsafe code in async_graph or smol

**Comment by tmiasko** [comment]
There is a single and incorrect use of unsafe in async_graph:

```
$ env RUSTFLAGS=-Zsanitizer=memory cargo -Zbuild-std run --target x86_64-unknown-linux-gnu 
...
    Finished dev [unoptimized + debuginfo] target(s) in 0.10s
     Running `target/x86_64-unknown-linux-gnu/debug/issue_minimal`
Works: 
==27447==WARNING: MemorySanitizer: use-of-uninitialized-value
...
    #2 0x55bb619da926 in async_graph::splitter::_$LT$impl$u20$async_graph..Splitter$u20$for$u20$async_graph..SplitRec$LT$A$GT$$GT$::split::h4b9f7e0308aa3072 /../.cargo/git/checkouts/asyncgraph-9a60b9185121d673/e8ef35f/async_graph/src/splitter.rs:90:5
...
SUMMARY: MemorySanitizer: use-of-uninitialized-value
Exiting
```

**Comment by SiebenCorgie** [comment]
Okay that seams to be the reason then, sorry for the inconvinience. I thought I'm sure that the bug originates outside of async_graph:

Thanks!
